packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/assign-public-ip/lambda/lib/route53.py

```diff
@@ -0,0 +1,91 @@
+from dataclasses import dataclass
+from typing import *
+import logging
+
+
+@dataclass
+class Route53RecordSetLocator:
+    hosted_zone_id: str
+    record_name: str
+
+
+# TODO: NEEDS OVERHAUL
+class Route53RecordSetMerger:
+    route53_client: Any
+    dynamo_table_client: Any
+
+    def __init__(self, route53_client):
+        self.route53_client = route53_client
+
+    def merge(self, locator: Route53RecordSetLocator, adds: List[str] = None, removes: List[str] = None):
+        adds = [] if adds is None else adds
+        removes = [] if removes is None else removes
+
+        record_set, is_new = self.get_record_set(locator)
+        new_record_set = merge_record_set_changes(record_set=record_set, adds=adds, removes=removes)
+
+        if len(new_record_set['ResourceRecords']) > 0:
+            self.upsert(locator, new_record_set)
+        elif not is_new:
+            self.delete(locator, record_set)
+        else:
+            logging.warning('Refusing to do anything with an new but empty recordset')
+
+    def get_record_set(self, locator: Route53RecordSetLocator):
+        record_type = 'A'
+        result = self.route53_client.list_resource_record_sets(HostedZoneId=locator.hosted_zone_id,
+                                                               StartRecordName=locator.record_name,
+                                                               StartRecordType=record_type, MaxItems="1")
+
+        record_sets = result['ResourceRecordSets']
+        if len(record_sets) > 0:
+            logging.info(f'Found a pre-existing record set: {record_sets[0]}')
+            return record_sets[0], False
+
+        logging.info('Creating a new record set')
+        return {'Name': locator.record_name, 'Type': record_type, 'ResourceRecords': [], 'TTL': 60}, True
+
+    def upsert(self, locator: Route53RecordSetLocator, record_set):
+        logging.info(f'Upserting record set {record_set}')
+        self.route53_client.change_resource_record_sets(
```

Comment:
I believe we need an error catch and retry here. According to the docs:

> ChangeResourceRecordSets requests
If Route 53 can't process a request before the next request arrives, it will reject subsequent requests for the same hosted zone and return an HTTP 400 error (Bad request). The response header also includes a Code element with a value of PriorRequestNotComplete and a Message element with a value of The request was rejected because Route 53 was still processing a prior request.

https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/DNSLimitations.html#limits-api-requests

Reply:
@nathanpeck Definitely worth considering. I have to perform an overhaul of this class anyway since it's not yet hooked up to the DynamoDB table info.

Otherwise, as it relates to Route53, I've been worrying a little about the potential for another race condition. It seems to me that the part that happens after the DynamoDB item is merged/received is still vulnerable to a data race. Diagram below. Given this, I wonder if I should stick with reserved concurrent executions of 1 for the lambda in question.

**Race Diagram**
Event A comes in followed by Event B. The optimistic locking for the DynamoDB portion results in both A's DDB Record and B's DDB Record being correct, even though Event A and event B happen concurrently. A's Record Set is out of date because event B happened, but it doesn't know it. Then, A and B are racing to put their version up on Route53. If A receives a delay, B may update the Record Set first, then A may overwrite B's Record Set without realizing that it's out of date.

```
Time ->

Event A -------> DDB Merge -------> A's DDB Record --------------> Backoff delay --------------> A's Record Set
     Event B -------> DDB Merge -------> B's DDB Record --------------> B's Record Set
```

Reply:
Yep you are totally right about that race. I think a concurrency of 1 for the lambda is necessary for sure and would solve the problem. Events should just queue up in SQS and apply one at a time. In fact if multiple events had queued up while the first invocation was happening then the next batch would actually come through as multiple records instead of one record so that would help solve the problem of hitting rate limits

Reply:
I've added the catch you mentioned in a recent commit, have added the retries by creating a helper, and dealt with the race condition we discussed by dropping to concurrency of 1.